buildSrc/src/main/java/com/google/firebase/gradle/ReleaseGenerator.kt

```diff
@@ -0,0 +1,79 @@
+package com.google.firebase.gradle
+
+import java.io.File
+import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.api.ListBranchCommand
+import org.eclipse.jgit.api.errors.GitAPIException
+import org.eclipse.jgit.lib.Constants
+import org.eclipse.jgit.lib.ObjectId
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+
+data class FirebaseLibrary(val moduleNames: List<String>, val directories: List<String>)
+
+open class ReleaseGenerator : DefaultTask() {
+    @TaskAction
+    @Throws(Exception::class)
+    fun generateReleaseConfig() {
+        val currentRelease = project.property("currentRelease").toString()
+        val pastRelease = project.property("pastRelease").toString()
+        val rootDir = project.rootDir
+        val availableModules = parseSubProjects(rootDir)
```

Comment:
Getting the candidates by parsing the `subprojects.cfg` file itself can introduce subtle bugs. Not all subprojects listed there are actually "firebase libraries". For example `firebase-config:bandwagoner` which is only used for testing. You can filter the projects themselves based on whether they have the firebase library extension like this

```
        val firebaseLibraries = project.subprojects
                .mapNotNull { it.extensions.findByType(FirebaseLibraryExtension::class.java) }
                .filterNot { it.artifactId.get().contains("ktx") }
                .map { e ->
                    val librariesPathToRelease = e.librariesToRelease.map { it.project.path.drop(1) }
                    FirebaseLibrary(
                            librariesPathToRelease,
                            librariesPathToRelease.map { it.replace(":", "/") })
                }
```

Reply:
`integ-testing` lives in `subprojects.cfg` and has a `gradle.properties` file but doesn't release.